pennylane_lightning/src/simulators/lightning_kokkos/algorithms/AdjointJacobianKokkos.hpp

```diff
@@ -0,0 +1,337 @@
+#pragma once
+#include "AdjointJacobianBase.hpp"
+#include "ObservablesKokkos.hpp"
+
+/// @cond DEV
+namespace {
+using namespace Pennylane::LightningKokkos::Observables;
+using namespace Pennylane::Algorithms;
+using Pennylane::LightningKokkos::Util::getImagOfComplexInnerProduct;
+} // namespace
+/// @endcond
+
+namespace Pennylane::LightningKokkos::Algorithms {
+
+/**
+ * @brief Kokkos-enabled adjoint Jacobian evaluator following the method of
+ * arXiV:2009.02823
+ *
+ * @tparam StateVectorT State vector type.
+ */
+template <class StateVectorT>
+class AdjointJacobian final
+    : public AdjointJacobianBase<StateVectorT, AdjointJacobian<StateVectorT>> {
+  private:
+    using ComplexT = typename StateVectorT::ComplexT;
+    using PrecisionT = typename StateVectorT::PrecisionT;
+    using BaseType =
+        AdjointJacobianBase<StateVectorT, AdjointJacobian<StateVectorT>>;
+
+    /**
+     * @brief Utility method to update the Jacobian at a given index by
+     * calculating the overlap between two given states.
+     *
+     * @param sv1 Statevector <sv1|. Data will be conjugated.
+     * @param sv2 Statevector |sv2>
+     * @param jac Jacobian receiving the values.
+     * @param scaling_coeff Generator coefficient for given gate derivative.
+     * @param obs_index Observable index position of Jacobian to update.
+     * @param param_index Parameter index position of Jacobian to update.
+     */
+    inline void updateJacobian(StateVectorT &sv1, StateVectorT &sv2,
+                               std::vector<std::vector<PrecisionT>> &jac,
+                               PrecisionT scaling_coeff, size_t obs_index,
+                               size_t param_index) {
+        jac[obs_index][param_index] = -2 * scaling_coeff *
+                                      getImagOfComplexInnerProduct<PrecisionT>(
+                                          sv1.getData(), sv2.getData());
+    }
+
+    /**
+     * @brief Utility method to apply all operations from given
+     * `%OpsData<StateVectorT>` object to
+     * `%StateVectorT`
+     *
+     * @param state Statevector to be updated.
+     * @param operations Operations to apply.
+     * @param adj Take the adjoint of the given operations.
+     */
+    inline void applyOperations(StateVectorT &state,
+                                const OpsData<StateVectorT> &operations,
+                                bool adj = false) {
+        for (size_t op_idx = 0; op_idx < operations.getOpsName().size();
+             op_idx++) {
+            state.applyOperation(operations.getOpsName()[op_idx],
+                                 operations.getOpsWires()[op_idx],
+                                 operations.getOpsInverses()[op_idx] ^ adj,
+                                 operations.getOpsParams()[op_idx]);
+        }
+    }
```

Comment:
Please check the AdjointJacobianBase class. Some of this agnostic functionality is already defined there.

Reply:
Saw that. There are also some rather device agnostic functionality in AdjointJacobian(Kokkos) which I moved in AdjointJacobianBase.